### ISSUE TITLE
A g:fzf_launcher possibly resulting in a more nicely-looking window

### ISSUE THREAD

**Issue by lydell** [junegunn/fzf#212]
The following comes from my .vimrc and it works well for me. Note that I had never used xterm before installing fzf—I even didn’t know that it was installed on my system! So there is a risk that the following might conflict with a user’s xterm configuration. I don’t know.

```vim
let font = split(getfontname(), ' ')
let font_family = join(font[:-2], ' ')
let font_size = font[-1]
let bg = synIDattr(hlID("Normal"), "bg")
let fg = synIDattr(hlID("Normal"), "fg")
let g:fzf_launcher = 'xterm -T fzf'
    \ .' -bg "\'.bg.'" -fg "\'.fg.'"'
    \ .' -fa "'.font_family.'" -fs '.font_size
    \ .' -e bash -ic %s'
```

- Sets the title to “fzf” instead of “bash”.
- Uses the same background and foreground colors as your vim theme. (Default for me was black text on white background.)
- Uses the same font family and font size as is used in vim. (On my system, by default xterm tries to load some non-existing font and falls back on a super ugly one, and the default font size is a bit too small to be useful.)

Even better, if this could be interpolated before each run we’d get a nice placement of it too (and if you change theme or font during the session that will be reflected as well):

```vim
    \ .' -geometry '.&columns.'x'.&lines/2.'+'.getwinposx().'+'.getwinposy()
```

Do you think any of this could be used by default in fzf? Since it is all very general and just mimics the user’s own vim customizations I think it could work out.

If not, I’d like to ask for support to pass a function to fzf to produce the launcher command. Then I’ll also add this to the wiki.

Edit: Here’s a more readable version:

```vim
let font = split(getfontname(), ' ')
let g:fzf_launcher = printf('xterm -T fzf'
      \ .' -bg "\%s" -fg "\%s"'
      \ .' -fa "%s" -fs %d'
      \ .' -geometry %dx%d+%d+%d -e bash -ic %%s',
      \ synIDattr(hlID("Normal"), "bg"), synIDattr(hlID("Normal"), "fg"),
      \ join(font[:-2], ' '), font[-1],
      \ &columns, &lines/2, getwinposx(), getwinposy())
```

**Comment by junegunn** [comment]
Thanks, that looks pretty cool. Actually since I'm always on terminal, I haven't been able to pay enough attention to this "launcher" thing.

I tried your solution on MacVim and noticed that I had to change it a little to extract the right font size as on MacVim `getfontname()` returns something like `Menlo Regular:h11`. So I had to split it with `[: ]` and strip `h` from the last part. And even when the number is correctly extracted, the font on xterm looks larger than that on MacVim. I don't know why but since I don't use GVim, I didn't further look into it :)

Although I like what you did here, providing the right default that works on different platforms is a hard thing to do, so I think we should just allow function references as `launcher` option and put you solution on the wiki page.

However, there's an annoying issue with Funcrefs as `g:fzf_launcher`.

```vim
function! Foo()
  " ...
  return 'xxx'
endfunction
let g:fzf_launcher = function('Foo')
```

And we get `E704: Funcref variable name must start with a capital: g:fzf_launcher`. So we may have to add another variable `g:Fzf_launcher`. (We can pass a funcref directly to `fzf#run()` function as in `call fzf#run({'launcher': function('Foo')})`, but then we'll have to pass it to every call to `fzf#run()`.) I'll let you know of the progress.

By the way, did you find a way to refocus Vim window after fzf is closed? On OS X, it is [possible with AppleScript](https://github.com/junegunn/fzf/wiki/On-MacVim-with-iTerm2) (a nice trick by @gleachkr), but I'm not sure if there's a similar trick that works on different platforms.

**Comment by junegunn** [comment]
You can use `g:Fzf_launcher` now.

**Comment by lydell** [comment]
> I had to change it a little to extract the right font size as on MacVim getfontname() returns something like Menlo Regular:h11. So I had to split it with [: ] and strip h from the last part.

`:help getfontname()` says that the function in question returns the font used by the `hl-Normal` highlighting group. `:help highlight-font` tells that this is platform dependent and that it is potentially very difficult to parse on X11.

Moreover, it seems like as long as I include `-fa "$x"` where `$x` is anything but the empty string, the system default monospace font is used, which is also the font I use in gvim. So that just happened to work for me.

There is also an `-fn` switch that seems to accept the name of a font family, but no matter what I pass in xterm always that it could not be found.

> And even when the number is correctly extracted, the font on xterm looks larger than that on MacVim.

`-fs 10` seems to work as expected for me though: A lower number makes the text smaller, and a larger number makes it bigger. `-fs 10` makes it exactly the same size as in gvim. Don’t know why it didn’t work for you.

> Although I like what you did here, providing the right default that works on different platforms is a hard thing to do, so I think we should just allow function references as launcher option and put you solution on the wiki page.

Yes, your tests in MacVim and the start of this comment clearly shows that—at least for the font stuff. But what about the `-T fzf`, the colors and the sizing?

- `-T fzf` seems harmless, and actually helps when you have many terminal windows open.
- Colors + sizing makes it feel like fzf was opened in a split at the top of vim, which many have requested.

In other words, I propose the following as a default `g:Fzf_launcher`:

```vim
function! Fzf_launcher()
  return printf('xterm -T fzf'
      \ .' -bg "\%s" -fg "\%s"'
      \ .' -geometry %dx%d+%d+%d -e bash -ic %%s',
      \ synIDattr(hlID("Normal"), "bg"), synIDattr(hlID("Normal"), "fg"),
      \ &columns, &lines/2, getwinposx(), getwinposy())
endfunction
```

When we have settled on this I’ll look into adding tips to the wiki.

> By the way, did you find a way to refocus Vim window after fzf is closed?

On Ubuntu and Trisquel gvim is refocused automatically by default when fzf is closed.

> You can use g:Fzf_launcher now.

Great! Thanks!

**Comment by junegunn** [comment]
I updated the default launcher as you proposed. Thanks!
